src/nl/rubensten/texifyidea/folding/LatexSectionFoldingBuilder.kt

```diff
@@ -4,72 +4,73 @@ import com.intellij.lang.ASTNode
 import com.intellij.lang.folding.FoldingBuilderEx
 import com.intellij.lang.folding.FoldingDescriptor
 import com.intellij.openapi.editor.Document
+import com.intellij.openapi.util.TextRange
 import com.intellij.psi.PsiElement
 import nl.rubensten.texifyidea.index.LatexCommandsIndex
-import nl.rubensten.texifyidea.psi.LatexCommands
 import nl.rubensten.texifyidea.psi.LatexContent
-import nl.rubensten.texifyidea.psi.PsiContainer
 import nl.rubensten.texifyidea.util.nextSiblingIgnoreWhitespace
 import nl.rubensten.texifyidea.util.parentOfType
 import nl.rubensten.texifyidea.util.previousSiblingIgnoreWhitespace
 
 /**
- * Folds multiple \\usepackage or \\RequirePackage statements
- *
- * @author Ruben Schellekens
+ * Recursively folds section commands
+ * @author Tom Evers
  */
 open class LatexSectionFoldingBuilder : FoldingBuilderEx() {
-
     companion object {
-
-        private val sectionCommands = arrayOf("\\section", "\\chapter")
+        private val sectionCommands = arrayOf("\\part", "\\chapter",
+                "\\section", "\\subsection", "\\subsubsection",
+                "\\paragraph", "\\subparagraph")
     }
 
     override fun isCollapsedByDefault(node: ASTNode) = false
 
     override fun getPlaceholderText(node: ASTNode) = node.text + "..."
 
-    override fun buildFoldRegions(root: PsiElement, document: Document, quick: Boolean): Array<FoldingDescriptor> {
+    override fun buildFoldRegions(root: PsiElement, document: Document,
+                                  quick: Boolean): Array<FoldingDescriptor> {
         val descriptors = ArrayList<FoldingDescriptor>()
         val commands = LatexCommandsIndex.getCommandsByNames(root.containingFile, *sectionCommands).toList()
-
+                .sortedBy { x -> x.textOffset }
         if (commands.isEmpty()) {
             return descriptors.toTypedArray()
         }
 
-        // Fold all sections that end with another section.
-        var previous = commands.first()
-        for (command in commands) {
-            if (previous == command) {
-                continue
+        for (index_cur in 0 until commands.size) {
```

Comment:
Rename `index_cur` to camelCase. Also what does 'cur' mean? Is it 'current'? If so, consider renaming it to `currentIndex`.

_Even then... current index of what?_

---------

Renaming to camelCase also holds for all other variables/values.

---------

Also consider using more idiomatic Kotlin:
```kotlin
for (index_cur in commands.indices)
```